### ISSUE TITLE
HDF5 unit test failures with gcc 5.x and clang 3.7

### ISSUE THREAD

**Issue by rwcarlsen** [cyclus/cyclus#1232]
Here is a link to the failures: https://gist.github.com/rwcarlsen/3c35df49b8db6a7de0b5

Everything compiles fine, but the failures occur consistently at runtime.  They are observable on Arch, Ubuntu, and Mac with the newer versions of both gcc and clang.  Not sure if it's HDF5's problem or a problem with the way we are using HDF5.  They are deterministic and occur every time.  Integration tests that use HDF5 also fail, and I can't successfully run any simulation with the HDF5 backend with the newer compilers.

@scopatz - any ideas?

[edit] updated gist link

**Comment by rwcarlsen** [comment]
@scopatz - let me know as soon as you can if you think there might be a quick fix or not.  If not, we might want to consider moving this to 1.5 and publishing it as a known bug.

**Comment by scopatz** [comment]
Yup, I'll look at it this week.

**Comment by scopatz** [comment]
PS your gist @rwcarlsen still does not work.

**Comment by scopatz** [comment]
Which, I suppose, you could call a linker error :)

**Comment by scopatz** [comment]
My linker errors seem to be related to boost, coin, and xml++ but *not* HDF5

```
[ 49%] Linking CXX executable ../bin/cyclus
CMakeFiles/cyclus_cli.dir/cyclus.cc.o: In function `boost::program_options::validation_error::validation_error(boost::program_options::validation_error::kind_t, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int)':
cyclus.cc:(.text._ZN5boost15program_options16validation_errorC2ENS1_6kind_tERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESA_i[_ZN5boost15program_options16validation_errorC5ENS1_6kind_tERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESA_i]+0x3d): undefined reference to `boost::program_options::validation_error::get_template[abi:cxx11](boost::program_options::validation_error::kind_t)'
cyclus.cc:(.text._ZN5boost15program_options16validation_errorC2ENS1_6kind_tERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESA_i[_ZN5boost15program_options16validation_errorC5ENS1_6kind_tERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESA_i]+0x5b): undefined reference to `boost::program_options::error_with_option_name::error_with_option_name(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int)'
CMakeFiles/cyclus_cli.dir/cyclus.cc.o: In function `boost::program_options::variables_map::operator[](std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const':
cyclus.cc:(.text._ZNK5boost15program_options13variables_mapixERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE[_ZNK5boost15program_options13variables_mapixERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE]+0x1f): undefined reference to `boost::program_options::abstract_variables_map::operator[](std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const'
CMakeFiles/cyclus_cli.dir/cyclus.cc.o: In function `boost::program_options::basic_command_line_parser<char>::basic_command_line_parser(int, char const* const*)':
cyclus.cc:(.text._ZN5boost15program_options25basic_command_line_parserIcEC2EiPKPKc[_ZN5boost15program_options25basic_command_line_parserIcEC5EiPKPKc]+0x81): undefined reference to `boost::program_options::detail::cmdline::cmdline(std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&)'
CMakeFiles/cyclus_cli.dir/cyclus.cc.o: In function `boost::program_options::basic_command_line_parser<char>::extra_parser(boost::function1<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&>)':
cyclus.cc:(.text._ZN5boost15program_options25basic_command_line_parserIcE12extra_parserENS_9function1ISt4pairINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESA_ERKSA_EE[_ZN5boost15program_options25basic_command_line_parserIcE12extra_parserENS_9function1ISt4pairINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESA_ERKSA_EE]+0x42): undefined reference to `boost::program_options::detail::cmdline::set_additional_parser(boost::function1<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&>)'
CMakeFiles/cyclus_cli.dir/cyclus.cc.o: In function `std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > boost::program_options::to_internal<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >(std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&)':
cyclus.cc:(.text._ZN5boost15program_options11to_internalINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEESt6vectorIS7_SaIS7_EERKS8_IT_SaISB_EE[_ZN5boost15program_options11to_internalINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEESt6vectorIS7_SaIS7_EERKS8_IT_SaISB_EE]+0x6c): undefined reference to `boost::program_options::to_internal(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
CMakeFiles/cyclus_cli.dir/cyclus.cc.o:(.rodata._ZTVN5boost15program_options11typed_valueIjcEE[_ZTVN5boost15program_options11typed_valueIjcEE]+0x38): undefined reference to `boost::program_options::value_semantic_codecvt_helper<char>::parse(boost::any&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, bool) const'
CMakeFiles/cyclus_cli.dir/cyclus.cc.o:(.rodata._ZTVN5boost15program_options11typed_valueINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEcEE[_ZTVN5boost15program_options11typed_valueINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEcEE]+0x38): undefined reference to `boost::program_options::value_semantic_codecvt_helper<char>::parse(boost::any&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, bool) const'
CMakeFiles/cyclus_cli.dir/cyclus.cc.o:(.rodata._ZTVN5boost15program_options20invalid_option_valueE[_ZTVN5boost15program_options20invalid_option_valueE]+0x30): undefined reference to `boost::program_options::error_with_option_name::substitute_placeholders(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const'
CMakeFiles/cyclus_cli.dir/cyclus.cc.o:(.rodata._ZTVN5boost15program_options16validation_errorE[_ZTVN5boost15program_options16validation_errorE]+0x30): undefined reference to `boost::program_options::error_with_option_name::substitute_placeholders(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const'
CMakeFiles/cyclus_cli.dir/cyclus.cc.o: In function `boost::program_options::typed_value<unsigned int, char>::name() const':
cyclus.cc:(.text._ZNK5boost15program_options11typed_valueIjcE4nameEv[_ZNK5boost15program_options11typed_valueIjcE4nameEv]+0x41): undefined reference to `boost::program_options::arg[abi:cxx11]'
CMakeFiles/cyclus_cli.dir/cyclus.cc.o: In function `boost::program_options::typed_value<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, char>::name() const':
cyclus.cc:(.text._ZNK5boost15program_options11typed_valueINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEcE4nameEv[_ZNK5boost15program_options11typed_valueINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEcE4nameEv]+0x41): undefined reference to `boost::program_options::arg[abi:cxx11]'
CMakeFiles/cyclus_cli.dir/cyclus.cc.o: In function `boost::program_options::typed_value<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, char>::xparse(boost::any&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&) const':
cyclus.cc:(.text._ZNK5boost15program_options11typed_valueINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEcE6xparseERNS_3anyERKSt6vectorIS7_SaIS7_EE[_ZNK5boost15program_options11typed_valueINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEcE6xparseERNS_3anyERKSt6vectorIS7_SaIS7_EE]+0xa8): undefined reference to `boost::program_options::validate(boost::any&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, int)'
CMakeFiles/cyclus_cli.dir/cyclus.cc.o: In function `void boost::program_options::validate<unsigned int, char>(boost::any&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, unsigned int*, long)':
cyclus.cc:(.text._ZN5boost15program_options8validateIjcEEvRNS_3anyERKSt6vectorINSt7__cxx1112basic_stringIT0_St11char_traitsIS7_ESaIS7_EEESaISB_EEPT_l[_ZN5boost15program_options8validateIjcEEvRNS_3anyERKSt6vectorINSt7__cxx1112basic_stringIT0_St11char_traitsIS7_ESaIS7_EEESaISB_EEPT_l]+0x115): undefined reference to `boost::program_options::invalid_option_value::invalid_option_value(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
CMakeFiles/cyclus_cli.dir/cyclus.cc.o:(.rodata._ZTVN5boost16exception_detail10clone_implINS0_19error_info_injectorINS_15program_options16validation_errorEEEEE[_ZTVN5boost16exception_detail10clone_implINS0_19error_info_injectorINS_15program_options16validation_errorEEEEE]+0x38): undefined reference to `boost::program_options::error_with_option_name::substitute_placeholders(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const'
CMakeFiles/cyclus_cli.dir/cyclus.cc.o:(.rodata._ZTVN5boost16exception_detail19error_info_injectorINS_15program_options16validation_errorEEE[_ZTVN5boost16exception_detail19error_info_injectorINS_15program_options16validation_errorEEE]+0x30): undefined reference to `boost::program_options::error_with_option_name::substitute_placeholders(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const'
CMakeFiles/cyclus_cli.dir/cyclus.cc.o:(.rodata._ZTVN5boost16exception_detail10clone_implINS0_19error_info_injectorINS_15program_options20invalid_option_valueEEEEE[_ZTVN5boost16exception_detail10clone_implINS0_19error_info_injectorINS_15program_options20invalid_option_valueEEEEE]+0x38): undefined reference to `boost::program_options::error_with_option_name::substitute_placeholders(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const'
CMakeFiles/cyclus_cli.dir/cyclus.cc.o:(.rodata._ZTVN5boost16exception_detail19error_info_injectorINS_15program_options20invalid_option_valueEEE[_ZTVN5boost16exception_detail19error_info_injectorINS_15program_options20invalid_option_valueEEE]+0x30): undefined reference to `boost::program_options::error_with_option_name::substitute_placeholders(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const'
../lib/libcyclus.so: undefined reference to `boost::archive::xml_oarchive_impl<boost::archive::xml_oarchive>::save(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
../lib/libcyclus.so: undefined reference to `OsiSolverInterface::addRow(CoinPackedVectorBase const&, double, double, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)'
../lib/libcyclus.so: undefined reference to `boost::archive::xml_iarchive_impl<boost::archive::xml_iarchive>::load(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&)'
../lib/libcyclus.so: undefined reference to `Glib::ustring::ustring(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
../lib/libcyclus.so: undefined reference to `OsiSolverInterface::addRow(CoinPackedVectorBase const&, char, double, double, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)'
../lib/libcyclus.so: undefined reference to `xmlpp::Node::get_children[abi:cxx11](Glib::ustring const&)'
../lib/libcyclus.so: undefined reference to `OsiSolverInterface::addCol(int, int const*, double const*, double, double, double, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)'
../lib/libcyclus.so: undefined reference to `xmlpp::Node::get_children[abi:cxx11](Glib::ustring const&) const'
../lib/libcyclus.so: undefined reference to `OsiSolverInterface::addCol(CoinPackedVectorBase const&, double, double, double, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)'
collect2: error: ld returned 1 exit status
cli/CMakeFiles/cyclus_cli.dir/build.make:168: recipe for target 'bin/cyclus' failed
make[2]: *** [bin/cyclus] Error 1
CMakeFiles/Makefile2:1493: recipe for target 'cli/CMakeFiles/cyclus_cli.dir/all' failed
make[1]: *** [cli/CMakeFiles/cyclus_cli.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
```

**Comment by scopatz** [comment]
Also, there a ton of compiler warnings (on Ubuntu 15.10 with GCC 5.2). These should probably be cleaned up.

**Comment by rwcarlsen** [comment]
I fixed the link - sorry for the delay.

**Comment by rwcarlsen** [comment]
Those linker errors are not something I've seen (at least w.r.t. this issue).  They don't appear to be particularly related to the hdf5 issues we've been seeing.

**Comment by scopatz** [comment]
Right, except that they are preventing me from getting to the HDF5 error or installing cyclus at all on my desktop.  Note that I am using miniconda3 with our stack on that machine.

**Comment by scopatz** [comment]
The errors that you posted for hdf5 seem like they are related to invalid memory locations for table and group locations.  My suspicion is that this is not an HDF5 version issue per se, or at least not only.  It should be possible to git bisect cyclus to find the change set that causes the problem.

**Comment by scopatz** [comment]
I would do this but I can't even compile the develop branch right now....

**Comment by rwcarlsen** [comment]
The issue may have existed forever in Cyclus - since it was only uncovered by new compiler versions :-(.  And I've still never figured out a good way to git bisect only --first-parent commits (i.e. excluding merge commits).

**Comment by scopatz** [comment]
I am happy to try to track down the issue, but I need to be able to compile cyclus.

**Comment by scopatz** [comment]
Also happy to open a separate issue for that.

**Comment by scopatz** [comment]
Any ideas?

**Comment by rwcarlsen** [comment]
I haven't had much chance to look.  My attention will be spotty as I
prepare for my defense over the next couple weeks.

**Comment by gonuke** [comment]
@scopatz can you get cyclus to build outside of a conda environment?  I can try if you haven't (notwithstanding #1236)

**Comment by rwcarlsen** [comment]
This is still an issue and will block the release until we can figure out a solution.  I dug at it a bit before and wasn't able to come up with anything. I'm not confident I'll be able to resolve it quickly - so continued help from others will be appreciated.

**Comment by scopatz** [comment]
I don't believe that this is - or at least not strictly - a compiler issue, but rather an HDF5 version issue.  @hodger and I went through a testing round and discovered that v1.8.13 was the last version that fully worked on Ubuntu 15.10 with the newer compilers.  We have yet to test the latest version (v1.8.16) to see if that addresses the problem.  We have been working on it. but more eyeballs would be nice.  Could you test if 1.8.13 works for you?

**Comment by rwcarlsen** [comment]
I get all the same failures with 1.8.13 with clang 3.7.1 on arch linux.

**Comment by rwcarlsen** [comment]
I ran the hdf5 library tests (i.e. `make test` in hdf5 source) and they all seem to pass when using a configuration that fails our Hdf5Back tests - increasing the likelyhood that the issue is in our code.

**Comment by scopatz** [comment]
Can you try HDF5 v1.8.16 please?

**Comment by rwcarlsen** [comment]
I think I've fixed the bug @scopatz - can you try #1232?

**Comment by rwcarlsen** [comment]
Sorry - I meant #1241
